// angular-cli/models/webpack-build-development.js
"use strict";
var path = require('path');
var webpack = require('webpack');
;
exports.getWebpackDevConfigPartial = function (projectRoot, appConfig) {
    return {
        devtool: 'source-map',
        output: {
            path: path.resolve(projectRoot, appConfig.outDir),
            filename: '[name].bundle.js',
            sourceMapFilename: '[name].map',
            chunkFilename: '[id].chunk.js'
        },
        plugins: [
            new webpack.LoaderOptionsPlugin({
                options: {
                    tslint: {
                        emitErrors: false,
                        failOnHint: false,
                        resourcePath: path.resolve(projectRoot, appConfig.root)
                    },
                }
            })
        ],
        node: {
            fs: 'empty',
            global: true,
            crypto: 'empty',
            process: true,
            module: false,
            clearImmediate: false,
            setImmediate: false
        }
    };
};
